File: 지은/Lv.2/방문길이.py
def solution(dirs):
    visited = set()
    x, y = 0, 0

    move = {"U": (0, 1), "D": (0, -1), "R": (1, 0), "L": (-1, 0)}

    for dir in dirs:
        dx, dy = move[dir]
        nx, ny = x + dx, y + dy

        if -5 <= nx <= 5 and -5 <= ny <= 5:
            if ((x, y), (nx, ny)) not in visited:
                visited.add(((x, y), (nx, ny)))
                visited.add(((nx, ny), (x, y)))
            x, y = nx, ny

    return len(visited) / 2

File: 화연/Lv.2/방문길이.js
/**
 * https://school.programmers.co.kr/learn/courses/30/lessons/49994
 * @param {string} dirs
 * @returns {number}
 */
function solution(dirs) {
  const dirObj = { U: 1, D: -1, R: 1, L: -1 };
  const root = [];
  let posi = [5, 5];
  for (const dir of [...dirs]) {
    let [prevx, prevy] = posi;
    if (dir === "U") {
      if (posi[0] === 10) continue;
      posi[0] += 1;
    } else if (dir === "D") {
      if (posi[0] === 0) continue;
      posi[0] -= 1;
    } else if (dir === "R") {
      if (posi[1] === 10) continue;
      posi[1] += 1;
    } else {
      if (posi[1] === 0) continue;
      posi[1] -= 1;
    }
    const move = `(${prevx}, ${prevy})-(${posi[0]}, ${posi[1]})`;
    const move2 = `(${posi[0]}, ${posi[1]})-(${prevx}, ${prevy})`;
    root.push(move);
    root.push(move2);
  }
  return new Set(root).size / 2;
}
